lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -101,29 +101,59 @@ class SimplifyReduceCvt : public mlir::RewritePattern {
     auto convert = llvm::cast<triton::gpu::ConvertLayoutOp>(op);
     triton::ReduceOp reduce;
     for (auto &use : convert.getResult().getUses()) {
-      auto owner = use.getOwner();
-      if (llvm::isa_and_nonnull<triton::ReduceOp>(owner)) {
-        reduce = llvm::cast<triton::ReduceOp>(owner);
-        break;
+      auto owner = llvm::dyn_cast<triton::ReduceOp>(use.getOwner());
+      if (!owner) {
+        continue;
+      }
+
+      // TODO: This always takes layout from the first argument which
+      // is fine for argmin/argmax but may not be optimal generally
```

Comment:
I think you limit all arguments of reduce to have the same encoding. So this is just fine?

>       if (t.getShape() != srcShape) {
>         rop.emitError() << "shape mismatch";
>       }

Reply:
The concern is that the first argument might be cheap to convert but the second argument slow to convert. In that case this will remove the cheap layout conversion and add a more expensive one.

Also, I don't think there's ever a case where shape mismatch can happen.